File: mac/agent.py
#!/usr/bin/env python3
"""Mac agent — long-poll Azure Service Bus and send via Messages.app.

Improvements in this version:
- Robust AMQP reconnect and credential refresh (recreate client on error)
- Exponential backoff with jitter on reconnect attempts
- Offload osascript sends to a short-lived worker thread and keep send/ack resolution in main thread
- Better logging and a simple health alert when repeated disconnects occur (optional http endpoint)

Auth: OAuth via DefaultAzureCredential (run `az login` once on the Mac).
Run with: uv run mac/agent.py
"""
import json
import logging
import os
import subprocess
import sys
import time
import threading
import queue
import random
from pathlib import Path
from typing import Dict, Any

from azure.identity import DefaultAzureCredential
from azure.servicebus import ServiceBusClient

# Optional: some SDKs raise these; we'll catch broadly and inspect when needed.
from azure.servicebus.exceptions import ServiceBusConnectionError, ServiceBusError

CFG_PATH = os.environ.get("IMSG_CONFIG", "config.json")
if not Path(CFG_PATH).exists():
    print(f"config not found: {CFG_PATH}", file=sys.stderr)
    sys.exit(2)

cfg = json.load(open(CFG_PATH))
FQDN = cfg["namespace_fqdn"]
QUEUE = cfg["queue"]
POLL = int(cfg.get("poll_interval_s", 3))
LOG_PATH = cfg.get("log_path", "./logs/agent.log")
HEALTH_ENDPOINT = cfg.get("health_endpoint")  # optional; POST when repeated disconnects

Path(LOG_PATH).parent.mkdir(parents=True, exist_ok=True)
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s %(levelname)s %(message)s",
    handlers=[logging.FileHandler(LOG_PATH), logging.StreamHandler()],
)
log = logging.getLogger("imsg-agent")

# Queues for offloading sends and returning results to main thread
send_queue: "queue.Queue[Dict[str, Any]]" = queue.Queue()
result_queue: "queue.Queue[Dict[str, Any]]" = queue.Queue()

# Mapping of internal_id -> (receiver, message)
pending_messages: Dict[str, Any] = {}
pending_lock = threading.Lock()

# Simple health counters
disconnect_count = 0
DISCONNECT_ALERT_THRESHOLD = int(cfg.get("disconnect_alert_threshold", 3))


def _osascript_send(to: str, body: str) -> bool:
    safe = body.replace("\\", "\\\\").replace('"', '\\"')
    script = (
        'tell application "Messages"\n'
        '  set targetService to 1st service whose service type = iMessage\n'
        f'  set theBuddy to buddy "{to}" of targetService\n'
        f'  send "{safe}" to theBuddy\n'
        "end tell"
    )
    try:
        # Keep capture_output to log detailed failures if they occur.
        subprocess.run(["osascript", "-e", script], check=True, capture_output=True, timeout=30)
        return True
    except subprocess.CalledProcessError as e:
        log.error("osascript failed for %s: %s", to, e.stderr.decode("utf-8", errors="replace"))
        return False
    except subprocess.TimeoutExpired:
        log.error("osascript timed out for %s", to)
        return False
    except Exception as e:
        log.exception("unexpected error running osascript for %s: %s", to, e)
        return False


def sender_worker():
    """Worker thread: take send jobs, run osascript, report results to result_queue."""
    while True:
        job = send_queue.get()
        if job is None:
            break
        internal_id = job["internal_id"]
        to = job["to"]
        body = job["body"]
        msg_id = job.get("msg_id", "?")

        success = _osascript_send(to, body)
        result_queue.put({"internal_id": internal_id, "success": success, "msg_id": msg_id})
        send_queue.task_done()


def post_health_alert(reason: str, details: str = ""):
    """If HEALTH_ENDPOINT is configured, POST a small alert JSON. Non-fatal if requests is missing."""
    if not HEALTH_ENDPOINT:
        return
    try:
        import requests

        payload = {"service": "imessage-mac-agent", "reason": reason, "details": details}
        requests.post(HEALTH_ENDPOINT, json=payload, timeout=5)
    except Exception as e:
        log.warning("failed to post health alert: %s", e)


def exponential_backoff(attempt: int, base: float = 1.0, cap: float = 60.0) -> float:
    """Compute exponential backoff with jitter."""
    exp = min(cap, base * (2 ** attempt))
    jitter = random.uniform(0, exp * 0.2)
    wait = exp + jitter
    return min(wait, cap)


def main() -> int:
    global disconnect_count

    log.info("starting agent — fqdn=%s queue=%s", FQDN, QUEUE)

    # Start a small sender worker thread. This keeps osascript off the receiver loop.
    worker = threading.Thread(target=sender_worker, daemon=True)
    worker.start()

    reconnect_attempt = 0
    while True:
        try:
            credential = DefaultAzureCredential()
            log.info("creating ServiceBusClient (attempt=%d)", reconnect_attempt)
            with ServiceBusClient(fully_qualified_namespace=FQDN, credential=credential) as client:
                reconnect_attempt = 0
                # Receiver context
                with client.get_queue_receiver(queue_name=QUEUE, max_wait_time=10) as receiver:
                    log.info("connected to service bus, listening...")
                    while True:
                        # First, handle any completed send results from worker(s)
                        try:
                            while True:
                                res = result_queue.get_nowait()
                                internal_id = res["internal_id"]
                                success = res["success"]
                                msg_id = res.get("msg_id", "?")
                                with pending_lock:
                                    entry = pending_messages.pop(internal_id, None)
                                if entry is None:
                                    log.warning("no pending message for %s", internal_id)
                                    continue
                                m = entry["message"]
                                # All SDK calls that mutate message state happen in main thread.
                                try:
                                    if success:
                                        receiver.complete_message(m)
                                        log.info("sent (async) %s", msg_id)
                                    else:
                                        receiver.abandon_message(m)
                                        log.warning("abandoned (async) %s for retry", msg_id)
                                except Exception as e:
                                    log.exception("failed to settle message %s: %s", msg_id, e)
                                result_queue.task_done()
                        except queue.Empty:
                            pass

                        try:
                            msgs = receiver.receive_messages(max_message_count=5, max_wait_time=10)
                        except ServiceBusConnectionError as e:
                            # Connection-level error: break to outer reconnect loop
                            log.exception("service bus connection error, will reconnect: %s", e)
                            raise
                        except Exception as e:
                            log.exception("unexpected receive error: %s", e)
                            # Treat as transient: break to recreate client
                            raise

                        if not msgs:
                            time.sleep(POLL)
                            continue

                        for m in msgs:
                            try:
                                payload = json.loads(b"".join(m.body).decode("utf-8"))
                            except Exception as e:
                                log.error("bad payload, dead-lettering: %s", e)
                                try:
                                    receiver.dead_letter_message(m, reason="bad-payload")
                                except Exception:
                                    log.exception("failed to dead-letter message")
                                continue

                            to = payload.get("to")
                            body = payload.get("body")
                            msg_id = payload.get("id", "?")

                            if not to or not body:
                                log.error("missing to/body in %s, dead-lettering", msg_id)
                                try:
                                    receiver.dead_letter_message(m, reason="missing-fields")
                                except Exception:
                                    log.exception("failed to dead-letter message")
                                continue

                            # Enqueue send job and track pending message
                            internal_id = f"{time.time()}-{random.randint(0,9999)}"
                            job = {"internal_id": internal_id, "to": to, "body": body, "msg_id": msg_id}
                            with pending_lock:
                                pending_messages[internal_id] = {"message": m, "enqueued_at": time.time()}
                            send_queue.put(job)
                            log.info("scheduled send %s -> %s (id=%s)", msg_id, to, internal_id)

        except KeyboardInterrupt:
            log.info("agent stopped by user")
            break
        except Exception as e:
            # Any exception here is treated as a broken connection: increment disconnect counter and backoff
            disconnect_count += 1
            log.exception("agent encountered error, will attempt reconnect: %s", e)
            if disconnect_count >= DISCONNECT_ALERT_THRESHOLD:
                log.warning("disconnect_count=%d exceeds threshold %d — posting health alert", disconnect_count, DISCONNECT_ALERT_THRESHOLD)
                post_health_alert("repeated-disconnects", details=f"count={disconnect_count}")
            wait = exponential_backoff(reconnect_attempt)
            log.info("waiting %.1fs before reconnect (attempt=%d)", wait, reconnect_attempt)
            time.sleep(wait)
            reconnect_attempt += 1
            continue

    # Clean up worker thread
    send_queue.put(None)
    worker.join(timeout=2)
    return 0


if __name__ == "__main__":
    try:
        sys.exit(main())
    except KeyboardInterrupt:
        log.info("agent stopped")
        sys.exit(0)

File: src/agent.mjs
/**
 * Mac agent — long-poll Service Bus and dispatch each message to Messages.app.
 *
 * Direct port of mac/agent.py. The Python version used a worker-thread queue
 * pattern because the SDK was sync-blocking; @azure/service-bus is fully
 * async/promise-based, so we just await each step inline. Same semantics:
 *   - peek-lock receive (durable until completed/abandoned)
 *   - dead-letter on bad payload or missing fields
 *   - complete on osascript success
 *   - abandon (re-deliverable) on osascript failure
 *   - exponential backoff with jitter on AMQP connection errors
 *   - optional health-alert webhook after N consecutive disconnects
 *
 * Auth: OAuth via DefaultAzureCredential (run `az login` once on the Mac).
 */

import { setTimeout as sleep } from "node:timers/promises";
import { DefaultAzureCredential } from "@azure/identity";
import { ServiceBusClient } from "@azure/service-bus";

import { createLogger } from "./log.mjs";
import { osascriptSend } from "./messages.mjs";

/**
 * @param {number} attempt  zero-based attempt counter
 * @param {{base?: number, capMs?: number}} [opts]
 * @returns {number} milliseconds to wait
 */
export function exponentialBackoff(attempt, { base = 1000, capMs = 60_000 } = {}) {
  const exp = Math.min(capMs, base * 2 ** attempt);
  const jitter = Math.random() * exp * 0.2;
  return Math.min(exp + jitter, capMs);
}

/**
 * Try to POST a small JSON alert to a webhook (config.health_endpoint).
 * Best-effort: never throws.
 *
 * @param {string|undefined} endpoint
 * @param {{reason: string, details?: string}} payload
 * @param {import("./log.mjs").Logger} [logger]
 */
export async function postHealthAlert(endpoint, payload, logger) {
  if (!endpoint) return;
  try {
    await fetch(endpoint, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({ service: "imessage-mac-agent", ...payload }),
      signal: AbortSignal.timeout(5000),
    });
  } catch (e) {
    logger?.warn(`failed to post health alert: ${e.message ?? e}`);
  }
}

/**
 * Process a single received message: parse → validate → osascript → settle.
 * Pure-ish — relies on injected receiver + sender so it's testable.
 *
 * @param {any} msg                       the ServiceBusReceivedMessage
 * @param {{
 *   receiver: any,
 *   sender: (to: string, body: string) => Promise<boolean>,
 *   logger: import("./log.mjs").Logger,
 * }} ctx
 */
export async function processOne(msg, { receiver, sender, logger }) {
  /** @type {{id?: string, to?: string, body?: string} | null} */
  let payload = null;
  try {
    const raw =
      typeof msg.body === "string"
        ? msg.body
        : Buffer.isBuffer(msg.body)
          ? msg.body.toString("utf8")
          : JSON.stringify(msg.body);
    payload = JSON.parse(raw);
  } catch (e) {
    logger.error(`bad payload, dead-lettering: ${e.message ?? e}`);
    try {
      await receiver.deadLetterMessage(msg, { deadLetterReason: "bad-payload" });
    } catch (settleErr) {
      logger.error(`failed to dead-letter: ${settleErr.message ?? settleErr}`);
    }
    return;
  }

  const { id = "?", to, body } = payload ?? {};
  if (!to || !body) {
    logger.error(`missing to/body in ${id}, dead-lettering`);
    try {
      await receiver.deadLetterMessage(msg, { deadLetterReason: "missing-fields" });
    } catch (settleErr) {
      logger.error(`failed to dead-letter: ${settleErr.message ?? settleErr}`);
    }
    return;
  }

  logger.info(`sending ${id} -> ${to}`);
  const ok = await sender(to, body);
  try {
    if (ok) {
      await receiver.completeMessage(msg);
      logger.info(`sent ${id}`);
    } else {
      await receiver.abandonMessage(msg);
      logger.warn(`abandoned ${id} for retry`);
    }
  } catch (settleErr) {
    logger.error(`failed to settle ${id}: ${settleErr.message ?? settleErr}`);
  }
}

/**
 * Run the agent. Blocks until SIGINT/SIGTERM or unrecoverable error.
 *
 * @param {{
 *   config: import("./config.mjs").BridgeConfig & {
 *     health_endpoint?: string,
 *     poll_interval_s?: number,
 *     disconnect_alert_threshold?: number,
 *     log_path?: string,
 *   },
 *   credential?: import("@azure/identity").TokenCredential,
 *   clientFactory?: (fqdn: string, cred: any) => ServiceBusClient,
 *   sender?: (to: string, body: string) => Promise<boolean>,
 *   logger?: import("./log.mjs").Logger,
 * }} opts
 * @returns {Promise<number>}  exit code
 */
export async function runAgent({ config, credential, clientFactory, sender, logger }) {
  const log = logger ?? createLogger(config.log_path ?? "./logs/agent.log");
  const fqdn = config.namespace_fqdn;
  const queue = config.queue;
  const pollMs = (config.poll_interval_s ?? 3) * 1000;
  const alertThreshold = config.disconnect_alert_threshold ?? 3;
  const send =
    sender ?? ((to, body) => osascriptSend({ to, body, logger: log }));

  log.info(`starting agent — fqdn=${fqdn} queue=${queue}`);

  let stop = false;
  const onSig = (sig) => {
    log.info(`received ${sig}, shutting down`);
    stop = true;
  };
  process.on("SIGINT", () => onSig("SIGINT"));
  process.on("SIGTERM", () => onSig("SIGTERM"));

  let disconnectCount = 0;
  let reconnectAttempt = 0;

  while (!stop) {
    /** @type {ServiceBusClient | null} */
    let client = null;
    try {
      const cred = credential ?? new DefaultAzureCredential();
      log.info(`creating ServiceBusClient (attempt=${reconnectAttempt})`);
      const make = clientFactory ?? ((f, c) => new ServiceBusClient(f, c));
      client = make(fqdn, cred);
      const receiver = client.createReceiver(queue, { receiveMode: "peekLock" });
      log.info("connected to service bus, listening...");
      reconnectAttempt = 0;

      try {
        while (!stop) {
          /** @type {any[]} */
          const msgs = await receiver.receiveMessages(5, { maxWaitTimeInMs: 10_000 });
          if (!msgs || msgs.length === 0) {
            await sleep(pollMs);
            continue;
          }
          // Process sequentially. osascript on Messages.app does NOT like
          // concurrent sends from the same script context; serial keeps
          // delivery order intact and avoids "buddy busy" failures.
          for (const m of msgs) {
            if (stop) break;
            await processOne(m, { receiver, sender: send, logger: log });
          }
        }
      } finally {
        try {
          await receiver.close();
        } catch (e) {
          log.warn(`receiver.close error: ${e.message ?? e}`);
        }
      }
    } catch (e) {
      disconnectCount += 1;
      log.error(`agent encountered error, will attempt reconnect: ${e.message ?? e}`);
      if (disconnectCount >= alertThreshold) {
        log.warn(
          `disconnect_count=${disconnectCount} exceeds threshold ${alertThreshold} — posting health alert`,
        );
        await postHealthAlert(
          config.health_endpoint,
          { reason: "repeated-disconnects", details: `count=${disconnectCount}` },
          log,
        );
      }
      const wait = exponentialBackoff(reconnectAttempt);
      log.info(`waiting ${(wait / 1000).toFixed(1)}s before reconnect (attempt=${reconnectAttempt})`);
      await sleep(wait);
      reconnectAttempt += 1;
    } finally {
      if (client) {
        try {
          await client.close();
        } catch (e) {
          log.warn(`client.close error: ${e.message ?? e}`);
        }
      }
    }
  }

  log.info("agent stopped");
  return 0;
}
